### ISSUE TITLE
torch.nn.functional.cosine_similarity got unexpected behavier with torch.int8 dtype

### ISSUE THREAD

**Issue by un-knight** [pytorch/pytorch#61454]
## 🐛 Bug

`torch.nn.functional.cosine_similarity` got unexpected behavier with torch.int8 dtype.

```
>>> a = torch.randn(10)*100
>>> a
tensor([ 47.8993, -27.2694,  20.9548, -13.5573, -54.0388, -76.8524,  -7.4037,
        -95.7477, 108.9276, -44.5625])
>>> b = a
>>> F.cosine_similarity(a, b, dim=-1)
tensor(1.)
>>> F.cosine_similarity(a.to(torch.int8), b, dim=-1)
tensor(3.5215e+12)
>>> F.cosine_similarity(a.to(torch.int16), b, dim=-1)
tensor(1.0000)
```

As we can see, `F.cosine_similarity(a.to(torch.int8), b, dim=-1)` got result of `tensor(3.5215e+12)`, while the expected result is `tensor(1.)`.

## Environment

PyTorch version: 1.9.0

cc @ezyang @gchanan @zou3519 @bdhirsh @jbschlosser @anjali411 @albanD @mruberry @nairbv

**Comment by albanD** [comment]
Isn't the problem that you downcast the value so badly (to int8) that the two numbers are now very different and so you get very different values?

**Comment by soulitzer** [comment]
I also tried:
```
>>> F.cosine_similarity(torch.tensor(12).to(torch.int8), torch.tensor(10.), dim=-1)
tensor(1.2000e+10)
>>> F.cosine_similarity(torch.tensor(10).to(torch.int8), torch.tensor(10.), dim=-1)
tensor(1.)
```

**Comment by albanD** [comment]
Indeed, converting back to float before calling the function does return the right value.
So something is wrong with the type promotion here.

**Comment by soulitzer** [comment]
@alban: two ints or two floats works

**Comment by nairbv** [comment]
A bit of investigation, the implementation of cosine_similarity is:
```
 Tensor cosine_similarity(const Tensor& x1, const Tensor& x2, int64_t dim, double eps) {
   // Follow scipy impl to improve numerical precision
   // Use x / sqrt(x * x) instead of x / (sqrt(x) * sqrt(x))
   Tensor w12 = at::sum(x1 * x2, dim);
   Tensor w1 = at::sum(x1 * x1, dim);
   Tensor w2 = at::sum(x2 * x2, dim);
   Tensor n12 = (w1 * w2).clamp_min_(eps * eps).sqrt_();
   return w12.div_(n12);
 }
```
what happens is the w2 value overflows:
```
In [23]: torch.tensor(12, dtype=torch.int8) * torch.tensor(12, dtype=torch.int8)
Out[23]: tensor(-112, dtype=torch.int8)
In [24]: 12 * 12
Out[24]: 144
In [25]: torch.iinfo(torch.int8)
Out[25]: iinfo(min=-128, max=127, dtype=int8)
```
To fix we can get a common dtype with `at::result_type(x1, x2);` and use that dtype when calculating w1, w2.

**Comment by nairbv** [comment]
FYI cosine_similarity type promotion is available due to the implementation relying on underlying operations that promote, and thus is probably undocumented.

Related: https://github.com/pytorch/pytorch/issues/56356

cc @mruberry
